src/Arcus.WebApi.Unit/Security/Authentication/SharedAccessKeyAuthenticationAttributeTests.cs

```diff
@@ -0,0 +1,150 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Arcus.Security.Secrets.Core.Interfaces;
+using Arcus.WebApi.Security.Authentication;
+using Arcus.WebApi.Unit.Hosting;
+using Microsoft.AspNetCore.Mvc;
+using Xunit;
+
+namespace Arcus.WebApi.Unit.Security.Authentication
+{
+    public class SharedAccessKeyAuthenticationAttributeTests : IDisposable
+    {
+        private const string HeaderName = "x-shared-access-key",
+                             SecretName = "custom-access-key-name",
+                             AuthorizedRoute = "/authz/shared-access-key";
+
+        private readonly TestApiServer _testServer = new TestApiServer();
+
+        [Theory]
+        [InlineData(null, "not empty or whitespace")]
+        [InlineData("", "not empty or whitespace")]
+        [InlineData(" ", "not empty or whitespace")]
+        [InlineData("not empty or whitespace", null)]
+        [InlineData("not empty or whitespace", "")]
+        [InlineData("not empty or whitespace", " ")]
+        public void SharedAccessKeyAttributeAndFilter_WithNotPresentHeaderNameAndOrSecretName_ShouldFailWithArgumentException(
+            string headerName,
+            string secretName)
+        {
+            Assert.Throws<ArgumentException>(
+                () => new SharedAccessKeyAuthenticationAttribute(headerName, secretName));
+
+            Assert.Throws<ArgumentException>(
+                () => new SharedAccessKeyAuthenticationFilter(headerName, secretName));
+        }
+
+        [Fact]
+        public async Task AuthorizedRoute_WithSharedAccessKey_ShouldFailWithKeyNotFoundException_WhenNoSecretProviderWasRegistered()
+        {
+            // Arrange
+            string secretValue = $"secret-{Guid.NewGuid()}";
+
+            // Act / Assert
+            await Assert.ThrowsAsync<KeyNotFoundException>(
+                () => SendAuthorizedHttpRequestWithHeader(HeaderName, secretValue));
+        }
+
+        [Fact]
+        public async Task AuthorizedRoute_WithSharedAccessKey_ShouldNotFailWithUnauthorized_WhenHeaderValueMatchesSecretValue()
+        {
+            // Arrange
+            string secretValue = $"secret-{Guid.NewGuid()}";
+            _testServer.AddService<ISecretProvider>(new InMemorySecretProvider((SecretName, secretValue)));
+
+            // Act
+            using (HttpResponseMessage response = await SendAuthorizedHttpRequestWithHeader(HeaderName, secretValue))
+            {
+                // Assert
+                Assert.NotEqual(HttpStatusCode.Unauthorized, response.StatusCode);
```

Comment:
I've misread this one, can't we assert on Ok or something instead?

Reply:
Well, it returned something likee `ContentTypeNotSupported` as status code. But I didn't quite found any reason (even when I explictily allowed all content-types).

But then, I thought about the what I wanted to test, and I don't want to test what status code the fixture (we ourselves created during the test setup) returns. But rather what the `SharedAccessKeyAuthenticationFIlter` NOT returned.

Therefore I find testing on NOT `Unauthorized` more correct.

Reply:
Fair enough!